tooling/nargo/src/ops/foreign_calls.rs

```diff
@@ -76,10 +126,92 @@ impl ForeignCall {
                     ],
                 })
             }
-            None => panic!("unexpected foreign call {foreign_call_name:?}"),
+            Some(ForeignCall::CreateMock) => {
+                let mock_oracle_name = Self::parse_string(&foreign_call.inputs[0]);
+                assert!(ForeignCall::lookup(&mock_oracle_name).is_none());
+                let id = self.last_mock_id;
+                self.mocked_responses.push(MockedCall::new(id, mock_oracle_name));
+                self.last_mock_id += 1;
+
+                Ok(ForeignCallResult { values: vec![Value::from(id).into()] })
+            }
+            Some(ForeignCall::SetMockParams) => {
+                let (id, params) = Self::extract_mock_id(&foreign_call.inputs)?;
+                self.find_mock_by_id(id)
+                    .unwrap_or_else(|| panic!("Unknown mock id {}", id))
+                    .params = Some(params.to_vec());
+
+                Ok(ForeignCallResult { values: vec![] })
+            }
+            Some(ForeignCall::SetMockReturns) => {
+                let (id, params) = Self::extract_mock_id(&foreign_call.inputs)?;
+                self.find_mock_by_id(id)
+                    .unwrap_or_else(|| panic!("Unknown mock id {}", id))
+                    .result = ForeignCallResult { values: params.to_vec() };
+
+                Ok(ForeignCallResult { values: vec![] })
+            }
+            Some(ForeignCall::SetMockTimes) => {
+                let (id, params) = Self::extract_mock_id(&foreign_call.inputs)?;
+                let times = params[0]
+                    .unwrap_value()
+                    .to_field()
+                    .try_to_u64()
+                    .expect("Invalid bit size of times");
+
+                self.find_mock_by_id(id)
+                    .unwrap_or_else(|| panic!("Unknown mock id {}", id))
+                    .times_left = Some(times);
+
+                Ok(ForeignCallResult { values: vec![] })
+            }
+            Some(ForeignCall::ClearMock) => {
+                let (id, _) = Self::extract_mock_id(&foreign_call.inputs)?;
+                self.mocked_responses.retain(|response| response.id != id);
+                Ok(ForeignCallResult { values: vec![] })
+            }
+            None => {
```

Comment:
So we are going to be assuming from this point on that any user-defined oracle is to be used as a mock oracle by `nargo`? I guess this is fine as any new oracles we would like to add will be need to be implemented in the foreign call executor with their own match case.

Reply:
This is for testing only, If we implement a solution for oracles in nargo execute, our solution should contact the custom resolver (inter-process communication or whatever solution we choose) in the case there are no mocks active :thinking:

Reply:
Yeah a full solution would require that. It looks like this is good to do as `nargo execute` is only used for testing and println, and if someone tries to define their own oracle to not used in a mock they will get an unknown foreing call error.